tests/k8st/tests/test_bgp_advert.py

```diff
@@ -319,6 +319,129 @@ def test_cluster_ip_advertisement(self):
+        """
+        Tests the node exclusion label.
+        - Create services, assert advertised from all nodes.
+        - Label one node so that it is excluded, assert that routes are withdrawn from that node.
+        - Delete / recreate service, assert it is still advertised from the correct nodes.
+        - Remove the exclusion label, assert that the node re-advertises the svc.
+        """
+        with DiagsCollector():
+
+            calicoctl("""apply -f - << EOF
+apiVersion: projectcalico.org/v3
+kind: BGPConfiguration
+metadata:
+  name: default
+spec:
+  serviceClusterIPs:
+  - cidr: 10.96.0.0/12
+  serviceExternalIPs:
+  - cidr: 175.200.0.0/16
+EOF
+""")
+
+            # Assert that a route to the service IP range is present.
+            cluster_cidr = "10.96.0.0/12"
+            retry_until_success(lambda: self.assertIn(cluster_cidr, self.get_routes()))
+
+            # Create both a Local and a Cluster type NodePort service with a single replica.
+            local_svc = "nginx-local"
+            cluster_svc = "nginx-cluster"
+            self.deploy("nginx:1.7.9", local_svc, self.ns, 80)
+            self.deploy("nginx:1.7.9", cluster_svc, self.ns, 80, traffic_policy="Cluster")
+            self.wait_until_exists(local_svc, "svc", self.ns)
+            self.wait_until_exists(cluster_svc, "svc", self.ns)
+
+            # Get clusterIPs.
+            local_svc_ip = self.get_svc_cluster_ip(local_svc, self.ns)
+            cluster_svc_ip = self.get_svc_cluster_ip(cluster_svc, self.ns)
+
+            # Wait for the deployments to roll out.
+            self.wait_for_deployment(local_svc, self.ns)
+            self.wait_for_deployment(cluster_svc, self.ns)
+
+            # Assert that both nginx service can be curled from the external node.
+            retry_until_success(curl, function_args=[local_svc_ip])
+            retry_until_success(curl, function_args=[cluster_svc_ip])
+
+            # Assert that local clusterIP is an advertised route and cluster clusterIP is not.
+            retry_until_success(lambda: self.assertIn(local_svc_ip, self.get_routes()))
+            retry_until_success(lambda: self.assertNotIn(cluster_svc_ip, self.get_routes()))
+
+            # Connectivity should always succeed.
+            for i in range(attempts):
+              retry_until_success(curl, function_args=[local_svc_ip])
+              retry_until_success(curl, function_args=[cluster_svc_ip])
+
+            # Scale local service to 4 replicas
+            self.scale_deployment(local_svc, self.ns, 4)
+            self.wait_for_deployment(local_svc, self.ns)
+            self.wait_for_deployment(cluster_svc, self.ns)
+
+            # Assert routes are correct and services are accessible.
+            # Local service should only be advertised from nodes that can run pods.
+            # The cluster CIDR should be advertised from all nodes.
+            self.assert_ecmp_routes(local_svc_ip, [self.ips[1], self.ips[2], self.ips[3]])
```

Comment:
we don't need to check `self.ips[0]` for `local_svc_ip`?

Reply:
nope, self.ips[0] is the controller node, which doesn't get pods scheduled to it.